File: atividades_posix/C/atividade4.c
#define _POSIX_C_SOURCE 200809L
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <pthread.h>
#include <time.h>

#define NUM_READERS 3
#define NUM_WRITERS 2
#define NUM_OPERATIONS 5

typedef struct {
    int data;
    pthread_rwlock_t rwlock;
} shared_resource_t;

shared_resource_t resource;

void* reader(void* arg) {
    int reader_id = *(int*)arg;
    
    for (int i = 0; i < NUM_OPERATIONS; i++) {
        pthread_rwlock_rdlock(&resource.rwlock);
        
        printf("Leitor %d: lendo dados = %d\n", reader_id, resource.data);
        usleep(rand() % 200000);
        
        pthread_rwlock_unlock(&resource.rwlock);
        
        usleep(rand() % 300000);
    }
    
    printf("Leitor %d: finalizando\n", reader_id);
    return NULL;
}

void* writer(void* arg) {
    int writer_id = *(int*)arg;
    
    for (int i = 0; i < NUM_OPERATIONS; i++) {
        int new_data = rand() % 1000 + 1;
        
        pthread_rwlock_wrlock(&resource.rwlock);
        
        printf("Escritor %d: escrevendo dados %d -> %d\n", 
               writer_id, resource.data, new_data);
        resource.data = new_data;
        usleep(rand() % 250000);
        
        pthread_rwlock_unlock(&resource.rwlock);
        
        usleep(rand() % 400000);
    }
    
    printf("Escritor %d: finalizando\n", writer_id);
    return NULL;
}

int main(int argc, char* argv[]) {
    int num_readers = NUM_READERS;
    int num_writers = NUM_WRITERS;
    
    if (argc == 3) {
        num_readers = atoi(argv[1]);
        num_writers = atoi(argv[2]);
    }
    
    printf("=== LEITORES/ESCRITORES - RWLOCKS (PRIORIDADE ESCRITORES) ===\n");
    printf("Operações por thread: %d\n", NUM_OPERATIONS);
    printf("Leitores: %d\n", num_readers);
    printf("Escritores: %d\n", num_writers);
    printf("=============================================================\n\n");
    
    srand(time(NULL));
    
    resource.data = 42;
    
    pthread_rwlockattr_t attr;
    pthread_rwlockattr_init(&attr);
    pthread_rwlockattr_setkind_np(&attr, PTHREAD_RWLOCK_PREFER_WRITER_NP);
    pthread_rwlock_init(&resource.rwlock, &attr);
    pthread_rwlockattr_destroy(&attr);
    
    pthread_t* reader_threads = malloc(num_readers * sizeof(pthread_t));
    pthread_t* writer_threads = malloc(num_writers * sizeof(pthread_t));
    int* reader_ids = malloc(num_readers * sizeof(int));
    int* writer_ids = malloc(num_writers * sizeof(int));
    
    for (int i = 0; i < num_writers; i++) {
        writer_ids[i] = i + 1;
        pthread_create(&writer_threads[i], NULL, writer, &writer_ids[i]);
    }
    
    for (int i = 0; i < num_readers; i++) {
        reader_ids[i] = i + 1;
        pthread_create(&reader_threads[i], NULL, reader, &reader_ids[i]);
    }
    
    for (int i = 0; i < num_writers; i++) {
        pthread_join(writer_threads[i], NULL);
    }
    
    for (int i = 0; i < num_readers; i++) {
        pthread_join(reader_threads[i], NULL);
    }
    
    printf("\n=== EXECUÇÃO FINALIZADA ===\n");
    printf("Valor final dos dados: %d\n", resource.data);
    
    pthread_rwlock_destroy(&resource.rwlock);
    
    free(reader_threads);
    free(writer_threads);
    free(reader_ids);
    free(writer_ids);
    
    return 0;
}

File: entrega1/atividades_posix/Java/Atividade4.java
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;
import java.util.Random;

class SharedResourceRW {
    private int data;
    private final ReadWriteLock rwLock = new ReentrantReadWriteLock(true);
    
    public SharedResourceRW(int initialData) {
        this.data = initialData;
    }
    
    public void read(int readerId) throws InterruptedException {
        rwLock.readLock().lock();
        try {
            System.out.printf("Leitor %d: lendo dados = %d%n", readerId, data);
            Thread.sleep(new Random().nextInt(200));
        } finally {
            rwLock.readLock().unlock();
        }
    }
    
    public void write(int writerId, int newData) throws InterruptedException {
        rwLock.writeLock().lock();
        try {
            System.out.printf("Escritor %d: escrevendo dados %d -> %d%n", 
                            writerId, data, newData);
            data = newData;
            Thread.sleep(new Random().nextInt(250));
        } finally {
            rwLock.writeLock().unlock();
        }
    }
    
    public int getData() {
        rwLock.readLock().lock();
        try {
            return data;
        } finally {
            rwLock.readLock().unlock();
        }
    }
}

class ReaderRW extends Thread {
    private final SharedResourceRW resource;
    private final int readerId;
    private final int numOperations;
    private final Random random = new Random();
    
    public ReaderRW(SharedResourceRW resource, int id, int numOperations) {
        this.resource = resource;
        this.readerId = id;
        this.numOperations = numOperations;
    }
    
    @Override
    public void run() {
        try {
            for (int i = 0; i < numOperations; i++) {
                resource.read(readerId);
                Thread.sleep(random.nextInt(300));
            }
            System.out.printf("Leitor %d: finalizando%n", readerId);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

class WriterRW extends Thread {
    private final SharedResourceRW resource;
    private final int writerId;
    private final int numOperations;
    private final Random random = new Random();
    
    public WriterRW(SharedResourceRW resource, int id, int numOperations) {
        this.resource = resource;
        this.writerId = id;
        this.numOperations = numOperations;
    }
    
    @Override
    public void run() {
        try {
            for (int i = 0; i < numOperations; i++) {
                int newData = random.nextInt(1000) + 1;
                resource.write(writerId, newData);
                Thread.sleep(random.nextInt(400));
            }
            System.out.printf("Escritor %d: finalizando%n", writerId);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

public class Atividade4 {
    private static final int NUM_OPERATIONS = 5;
    
    public static void main(String[] args) {
        int numReaders = 3;
        int numWriters = 2;
        
        if (args.length == 2) {
            numReaders = Integer.parseInt(args[0]);
            numWriters = Integer.parseInt(args[1]);
        }
        
        System.out.println("=== LEITORES/ESCRITORES - RWLOCKS (PRIORIDADE ESCRITORES) ===");
        System.out.println("Operações por thread: " + NUM_OPERATIONS);
        System.out.println("Leitores: " + numReaders);
        System.out.println("Escritores: " + numWriters);
        System.out.println("=============================================================\n");
        
        SharedResourceRW resource = new SharedResourceRW(42);
        
        WriterRW[] writers = new WriterRW[numWriters];
        ReaderRW[] readers = new ReaderRW[numReaders];
        
        for (int i = 0; i < numWriters; i++) {
            writers[i] = new WriterRW(resource, i + 1, NUM_OPERATIONS);
            writers[i].start();
        }
        
        for (int i = 0; i < numReaders; i++) {
            readers[i] = new ReaderRW(resource, i + 1, NUM_OPERATIONS);
            readers[i].start();
        }
        
        try {
            for (WriterRW writer : writers) {
                writer.join();
            }
            
            for (ReaderRW reader : readers) {
                reader.join();
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        
        System.out.println("\n=== EXECUÇÃO FINALIZADA ===");
        System.out.println("Valor final dos dados: " + resource.getData());
    }
}